### PR TITLE
GH-43953: [C++] Add tests based on random data and benchmarks to ChunkResolver::ResolveMany

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

Improve tests and add benchmarks. I wrote the tests and benchmarks while trying to improve the performance of `ResolveMany` and failing at it.

### What changes are included in this PR?

Tests, benchmarks, and changes that don't really affect performance but might unlock more optimization opportunities in the future.

### Are these changes tested?

Yes.

* GitHub Issue: #43953

### CODE REVIEW

**Comment by felipecrv** [comment]
The `chunk_hint` stuff really works. I forgot to plot it, but without it the throughput gains don't exist when indices are sorted.

resolvemany

**Comment by pitrou** [comment]
Results here (AMD Zen 2 CPU, gcc):
```
ResolveManyUInt16Random/65535/10000/508         7072 ns         7071 ns        98917 items_per_second=71.8471M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt16Random/65535/100/508           4098 ns         4097 ns       170804 items_per_second=123.986M/s logical_len=65.535k num_chunks=100
ResolveManyUInt16Random/65535/10000/32764    1812677 ns      1812178 ns          393 items_per_second=18.0799M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt16Random/65535/100/32764       723697 ns       723335 ns          876 items_per_second=45.2958M/s logical_len=65.535k num_chunks=100

ResolveManyUInt32Random/65535/10000/508         7796 ns         7795 ns        88954 items_per_second=65.1731M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt32Random/65535/100/508           4397 ns         4396 ns       157489 items_per_second=115.566M/s logical_len=65.535k num_chunks=100
ResolveManyUInt32Random/65535/10000/32764    1848926 ns      1848181 ns          379 items_per_second=17.7277M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt32Random/65535/100/32764       752273 ns       752045 ns          917 items_per_second=43.5665M/s logical_len=65.535k num_chunks=100

ResolveManyUInt64Random/65535/10000/508         6731 ns         6730 ns       103505 items_per_second=75.4878M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt64Random/65535/100/508           4042 ns         4041 ns       172940 items_per_second=125.717M/s logical_len=65.535k num_chunks=100
ResolveManyUInt64Random/65535/10000/32764    1809563 ns      1809068 ns          387 items_per_second=18.111M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt64Random/65535/100/32764       731165 ns       730968 ns          953 items_per_second=44.8227M/s logical_len=65.535k num_chunks=100

ResolveManyUInt16Sorted/65535/10000/508         6233 ns         6231 ns       111713 items_per_second=81.5236M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt16Sorted/65535/100/508            955 ns          955 ns       728588 items_per_second=532.031M/s logical_len=65.535k num_chunks=100
ResolveManyUInt16Sorted/65535/10000/32764     314016 ns       313931 ns         2221 items_per_second=104.367M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt16Sorted/65535/100/32764        41365 ns        41356 ns        17068 items_per_second=792.247M/s logical_len=65.535k num_chunks=100

ResolveManyUInt32Sorted/65535/10000/508         7176 ns         7174 ns        91879 items_per_second=70.8084M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt32Sorted/65535/100/508            928 ns          928 ns       749739 items_per_second=547.568M/s logical_len=65.535k num_chunks=100
ResolveManyUInt32Sorted/65535/10000/32764     326595 ns       326507 ns         2145 items_per_second=100.347M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt32Sorted/65535/100/32764        34162 ns        34154 ns        20518 items_per_second=959.309M/s logical_len=65.535k num_chunks=100

ResolveManyUInt64Sorted/65535/10000/508         6057 ns         6055 ns       113029 items_per_second=83.8968M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt64Sorted/65535/100/508            994 ns          993 ns       700153 items_per_second=511.41M/s logical_len=65.535k num_chunks=100
ResolveManyUInt64Sorted/65535/10000/32764     316107 ns       316023 ns         2212 items_per_second=103.676M/s logical_len=65.535k num_chunks=10k
ResolveManyUInt64Sorted/65535/100/32764        38441 ns        38427 ns        20599 items_per_second=852.64M/s logical_len=65.535k num_chunks=100
```

**Comment by pitrou** [comment]
Ironically, uint32 seems slightly slower than both uint16 and uint64. Not sure that's due to the compiler or to the CPU.

File: cpp/src/arrow/chunk_resolver_benchmark.cc [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    ARROW_CHECK(all_succeeded);
+    state.counters["logical_len"] = static_cast<double>(chunked_array_length);
+    state.counters["num_chunks"] = static_cast<double>(num_chunks);
+    state.SetItemsProcessed(state.iterations() * num_logical_indices);
+  }
+};
+
+template <typename IndexType>
+void ResolveManySetArgs(benchmark::internal::Benchmark* bench) {
+  constexpr int32_t kNonAligned = 3;
```
Comment (pitrou) [inline cpp/src/arrow/chunk_resolver_benchmark.cc:106]: Can you explain what this is for?

File: cpp/src/arrow/chunk_resolver_benchmark.cc [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  const int64_t kNumIndicesFew = (kChunkedArrayLength >> 7) - kNonAligned;
+  const int64_t kNumIndicesMany = (kChunkedArrayLength >> 1) - kNonAligned;
+
+  switch (sizeof(IndexType)) {
+    case 1:
+      // Unexpected. See comments below.
+    case 2:
+    case 4:
+    case 8:
+      bench->Args({kChunkedArrayLength, /*num_chunks*/ 10000, kNumIndicesFew});
```
Comment (pitrou) [inline cpp/src/arrow/chunk_resolver_benchmark.cc:116]: 10000 chunks is really a lot and I'm not sure it's really useful to test different numbers of chunks. By accumulating different combinations of parameters we make the benchmark results less immediately readable.

File: cpp/src/arrow/chunked_array.cc [hunk 3]
```diff
@@ -55,7 +55,7 @@ ChunkedArray::ChunkedArray(ArrayVector chunks, std::shared_ptr<DataType> type)
         << "cannot construct ChunkedArray from empty vector and omitted type";
     type_ = chunks_[0]->type();
   }
-
+  ARROW_CHECK_LE(chunks.size(), std::numeric_limits<int>::max());
```
Comment (pitrou) [inline cpp/src/arrow/chunked_array.cc:58]: Is this limit useful if it ends up not making performance better anyway?

File: cpp/src/arrow/chunked_array_test.cc [hunk 4]
```diff
@@ -373,10 +374,27 @@ TEST(TestChunkResolver, Resolve) {
...
+  std::sort(offsets.begin() + 1, offsets.end());
+  return offsets;
+}
+
 template <typename T>
 class TestChunkResolverMany : public ::testing::Test {
  public:
   using IndexType = T;
+  static constexpr int32_t kMaxInt32 = std::numeric_limits<int32_t>::max();
+  static constexpr size_t kMaxValidIndex = std::numeric_limits<IndexType>::max();
```
Comment (pitrou) [inline cpp/src/arrow/chunked_array_test.cc:397]: Shouldn't this be `uint64_t`? Especially for 32-bit platforms.

File: cpp/src/arrow/chunk_resolver_benchmark.cc (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/chunk_resolver_benchmark.cc:106]: The optimizations I was experimenting with involved some unrolling so I didn't want input values to neatly align to powers of 2.

File: cpp/src/arrow/chunked_array.cc (hunk 3 above)
Comment (felipecrv) [reply cpp/src/arrow/chunked_array.cc:58]: This is useful in ensuring we are not generating more chunks than can be addressed by the index type.

File: cpp/src/arrow/chunked_array_test.cc (hunk 4 above)
Comment (felipecrv) [reply cpp/src/arrow/chunked_array_test.cc:397]: True. Let me fix this.

File: cpp/src/arrow/chunk_resolver_benchmark.cc (hunk 2 above)
Comment (felipecrv) [reply cpp/src/arrow/chunk_resolver_benchmark.cc:116]: The huge number is necessary (even though it's a bit unrealistic) to measure how effective the binary search is at reducing the search space (proportional to the number of chunks).

**Comment by felipecrv** [comment]
> Ironically, uint32 seems slightly slower than both uint16 and uint64. Not sure that's due to the compiler or to the CPU.

Same on `12th Gen Intel(R) Core(TM) i7-12800H`. This is the reason I changed `lo` and `hi` to `uint32_t` and nothing changed in the benchmarks.

Things are saner on the Apple M1 Pro with 32-bit being the fastest.

```
Run on (8 X 24 MHz CPU s)
CPU Caches:
  L1 Data 64 KiB
  L1 Instruction 128 KiB
  L2 Unified 4096 KiB (x8)
Load Average: 14.23, 6.75, 4.28
------------------------------------------------------------------------------------------------
Benchmark                                      Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------
ResolveManyUInt16Random/65535/100/508       2769 ns         2747 ns       252543 items_per_second=184.959M/s logical_len=65.535k num_chunks=100
ResolveManyUInt32Random/65535/100/508       2715 ns         2689 ns       257431 items_per_second=188.93M/s logical_len=65.535k num_chunks=100
ResolveManyUInt64Random/65535/100/508       2928 ns         2757 ns       252188 items_per_second=184.281M/s logical_len=65.535k num_chunks=100
ResolveManyUInt16Sorted/65535/100/508       1023 ns         1012 ns       693694 items_per_second=501.751M/s logical_len=65.535k num_chunks=100
ResolveManyUInt32Sorted/65535/100/508        929 ns          920 ns       761052 items_per_second=551.881M/s logical_len=65.535k num_chunks=100
ResolveManyUInt64Sorted/65535/100/508        933 ns          926 ns       759878 items_per_second=548.86M/s logical_len=65.535k num_chunks=100
```

**Comment by pitrou** [comment]
There are a number of CI failures that need fixing.

**Comment by felipecrv** [comment]
> There are a number of CI failures that need fixing.

TIL `uniform_int_distribution` can't take `(u)int8_t` as type parameter on MSVC builds.

**Comment by pitrou** [comment]
There are still a couple CI failures that need fixing.

**Comment by pitrou** [comment]
There are still a couple compilation errors on CI it seems :-)

**Comment by felipecrv** [comment]
@pitrou now it's all green. The only failure is unrelated.

**Comment by pitrou** [comment]
Just a question: do you mean to keep the 64-bit to 32-bit chunk index change?

**Review by pitrou — APPROVED** [review]
LGTM after a minor push. I'll merge if CI is green.

**Comment by felipecrv** [comment]
> Just a question: do you mean to keep the 64-bit to 32-bit chunk index change?

Yes I do.

**Pull request merged**